### PR TITLE
comp_task requests are sequentially committed for each node in the project

### PR DESCRIPTION
**Pull request by GitHK**
## What do these changes do?

**NOTE: Original PR code was scraped as it did not address the issue**

# Issue explained by @sanderegg 

[see original below](https://github.com/ITISFoundation/osparc-simcore/pull/2043#issuecomment-743018422)

THE only one (service) inserting there is the director-v2. the sidecar/services are only updating.
Therefore, I think the problem is the following:

webclient -> PUT /projects/{uuid}
-> webserver saves project AND sends a /POST /computations/{uuid} to the director-v2
-> director-v2 inserts the new computation by deleting and then inserting
PROBLEM: if you move a node the webclient also triggers a PUT cause the node changed, therefore if the user moves his things/add nodes/modify nodes triggering a lot of PUTs, I think the director-v2 receives a string of requests (which are all async). so my guess is that the director-v2 is still treating the request when a nother arrives.

# Approved implementation proposal:

[see original below](https://github.com/ITISFoundation/osparc-simcore/pull/2043#issuecomment-743024764)
- I create something similar to the `run_sequentially_in_context` decorator which will be applied to a function responsible for updating each individual node in the project
- This will put calls to this function in a queue which will guarantee FIFO thus applying all the requests from the frontend in sequence.
- The delete recreate logic will still continue to exist and I will remove the upset patter to detect future errors.

## Checklist

- [ ] Did you change any service's API? Then make sure to bundle document and upgrade version (``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [ ] Unit tests for the changes exist
- [ ] Runs in the swarm
- [ ] Documentation reflects the changes
- [ ] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)

### CODE REVIEW

**Review by pcrespov — COMMENTED** [review]
this pattern is new to me. What is it about? Avoids the director-v2 getting angry? 😤 🤣 😉

File: services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_tasks.py [hunk 1]
```diff
@@ -132,11 +132,14 @@ async def upsert_tasks_from_project(
             )
             internal_id = internal_id + 1
 
-            await self.connection.execute(
-                insert(comp_tasks).values(
-                    **task_db.dict(by_alias=True, exclude_unset=True)
-                )
+            insert_stmt = insert(comp_tasks).values(
+                **task_db.dict(by_alias=True, exclude_unset=True)
```
Comment (pcrespov) [inline services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_tasks.py:136]: MINOR optimization: export to dict only once
```python

# ... let's upseRT (insert or update) so people do not get upset! ;-)

task_values: Dict[str,Any] = task_db.dict(by_alias=True, exclude_unset=True)
upsert_stmt = insert(comp_tasks).values(**task_values).on_conflict_do_update(
                index_elements=[comp_tasks.c.project_id, comp_tasks.c.node_id],
                set_=task_values,
            )
```

**Review by pcrespov — APPROVED** [review]
Some minor optimization ;-)

**Review by sanderegg — CHANGES_REQUESTED** [review]
sorry but I am against that change for now.
the first line in the function deletes the comp_tasks so the problem lies somewhere else.

there should be no need to upsert here.

**Comment by GitHK** [comment]
> sorry but I am against that change for now.
> the first line in the function deletes the comp_tasks so the problem lies somewhere else.
> 
> there should be no need to upsert here. the problem is probably more that this code gets called asynchronously several times. doing upserts here will certainly generate some undefined state in the DB.

You are correct about your assumption. I thought the same. But for some reason the another process/task is interacting with `comp_task` causing the item to already exist between the time you delete it and recreate it again, causing a duplicate key error. This happened yesterday on master. Have a look at the attached log (the error is still present in the deployment).

[unexpected_error.log](https://github.com/ITISFoundation/osparc-simcore/files/5677436/unexpected_error.log)

**Comment by sanderegg** [comment]
> > sorry but I am against that change for now.
> > the first line in the function deletes the comp_tasks so the problem lies somewhere else.
> > there should be no need to upsert here. the problem is probably more that this code gets called asynchronously several times. doing upserts here will certainly generate some undefined state in the DB.
> 
> You are correct about your assumption. I thought the same. But for some reason the another process/task is interacting with `comp_task` causing the item to already exist between the time you delete it and recreate it again, causing a duplicate key error. This happened yesterday on master.
> 
> Have a look at the attached log (the error is still present in the deployment).
> 
> [unexpected_error.log](https://github.com/ITISFoundation/osparc-simcore/files/5677436/unexpected_error.log)

yes. as I said. THE only one inserting there is the director-v2. the sidecar/services are only updating.
Therefore, I think the problem is the following:
- webclient -> PUT /projects/{uuid}
- -> webserver saves project AND sends a /POST /computations/{uuid} to the director-v2
- -> director-v2 inserts the new computation by deleting and then inserting
PROBLEM: if you move a node the webclient also triggers a PUT cause the node changed, therefore if the user moves his things/add nodes/modify nodes triggering a lot of PUTs, I think the director-v2 receives a string of requests (which are all async). so my guess is that the director-v2 is still treating the request when a nother arrives.

**Comment by GitHK** [comment]
> > > sorry but I am against that change for now.
> > > the first line in the function deletes the comp_tasks so the problem lies somewhere else.
> > > there should be no need to upsert here. the problem is probably more that this code gets called asynchronously several times. doing upserts here will certainly generate some undefined state in the DB.
> > 
> > 
> > You are correct about your assumption. I thought the same. But for some reason the another process/task is interacting with `comp_task` causing the item to already exist between the time you delete it and recreate it again, causing a duplicate key error. This happened yesterday on master.
> > Have a look at the attached log (the error is still present in the deployment).
> > [unexpected_error.log](https://github.com/ITISFoundation/osparc-simcore/files/5677436/unexpected_error.log)
> 
> yes. as I said. THE only one inserting there is the director-v2. the sidecar/services are only updating.
> Therefore, I think the problem is the following:
> 
> * webclient -> PUT /projects/{uuid}
> * -> webserver saves project AND sends a /POST /computations/{uuid} to the director-v2
> * -> director-v2 inserts the new computation by deleting and then inserting
>   PROBLEM: if you move a node the webclient also triggers a PUT cause the node changed, therefore if the user moves his things/add nodes/modify nodes triggering a lot of PUTs, I think the director-v2 receives a string of requests (which are all async). so my guess is that the director-v2 is still treating the request when a nother arrives.

That sounds really possible. Would you something like this working here:
- I create something similar to the`run_sequentially_in_context` decorator which will be applied to a function responsible for updating each individual node in the project
- This will put calls to this function in a queue which will guarantee FIFO thus applying all the requests from the frontend in sequence.
- The delete recreate logic will still continue to exist and I will remove the upset patter to detect future errors.

**Comment by sanderegg** [comment]
> > > > sorry but I am against that change for now.
> > > > the first line in the function deletes the comp_tasks so the problem lies somewhere else.
> > > > there should be no need to upsert here. the problem is probably more that this code gets called asynchronously several times. doing upserts here will certainly generate some undefined state in the DB.
> > > 
> > > 
> > > You are correct about your assumption. I thought the same. But for some reason the another process/task is interacting with `comp_task` causing the item to already exist between the time you delete it and recreate it again, causing a duplicate key error. This happened yesterday on master.
> > > Have a look at the attached log (the error is still present in the deployment).
> > > [unexpected_error.log](https://github.com/ITISFoundation/osparc-simcore/files/5677436/unexpected_error.log)
> > 
> > 
> > yes. as I said. THE only one inserting there is the director-v2. the sidecar/services are only updating.
> > Therefore, I think the problem is the following:
> > 
> > * webclient -> PUT /projects/{uuid}
> > * -> webserver saves project AND sends a /POST /computations/{uuid} to the director-v2
> > * -> director-v2 inserts the new computation by deleting and then inserting
> >   PROBLEM: if you move a node the webclient also triggers a PUT cause the node changed, therefore if the user moves his things/add nodes/modify nodes triggering a lot of PUTs, I think the director-v2 receives a string of requests (which are all async). so my guess is that the director-v2 is still treating the request when a nother arrives.
> 
> That sounds really possible. Would you something like this working here:
> 
> * I create something similar to the`run_sequentially_in_context` decorator which will be applied to a function responsible for updating each individual node in the project
> * This will put calls to this function in a queue which will guarantee FIFO thus applying all the requests from the frontend in sequence.
> * The delete recreate logic will still continue to exist and I will remove the upset patter to detect future errors.

this sounds like a good proposal.

File: services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_tasks.py [hunk 2]
```diff
@@ -73,6 +74,70 @@ async def get_comp_tasks(
...
+            internal_id=internal_id,
+            node_class=node_class,
+        )
+
+        await self.connection.execute(
+            insert(comp_tasks).values(**task_db.dict(by_alias=True, exclude_unset=True))
+        )
+
     @log_decorator(logger=logger)
     async def upsert_tasks_from_project(
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_tasks.py:142]: hmm I think the sequential decorator should be on that one don't you think?
what happens if we delete the tasks while the other is inserting?

**Review by sanderegg — CHANGES_REQUESTED** [review]

File: services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_tasks.py (hunk 2 above)
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_tasks.py:142]: I overlooked the delete. At this point the best thing to do is to lock the entire `upsert_tasks_from_project` based on `project_id`. Will guarantee that the delete recreate happens without external interventions.

Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_tasks.py:142]: 👍

File: services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_tasks.py [hunk 3]
```diff
@@ -103,9 +108,6 @@ async def upsert_tasks_from_project(
                 node_extras: ServiceExtras = await director_client.get_service_extras(
                     service_key_version
                 )
-            if not node_details:
-                continue
-
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_tasks.py:108]: why is this one gone? bad merge?

**Review by sanderegg — CHANGES_REQUESTED** [review]
see comment. I think something from master went away

File: services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_tasks.py (hunk 3 above)
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_tasks.py:108]: my fault was not able to merge it correctly. Will add it back.

**Review by sanderegg — APPROVED** [review]
all good!

**Pull request merged**